### PR TITLE
New anti-reset windup for multicopter position controller

### PR DESCRIPTION
**Pull request by bresch**
This PR introduces a "new" type of Anti-Reset Windup (ARW) method for integrator handling in the velocity controller. This type of method is knowm as "Tracking Anti-Windup" and is described in [this paper](https://ac.els-cdn.com/S1474667017518650/1-s2.0-S1474667017518650-main.pdf?_tid=ed135642-d163-4aa6-98e7-4a912a5a87a4&acdnat=1538986751_bb7f14c5bd5e1a5803f50659631e671d)
 
**Test data / coverage**
In the following log, the first step response is with no ARW, the 2nd one is with the legacy ARW and the 3rd one is with the new ARW strategy this PR introduces.
https://logs.px4.io/plot_app?log=b8a97c5f-d186-475f-a804-1eeff36e9866#Nav-Velocity
bokeh_plot 1

You can clearly see that in the first one, the integral takes time to slowly wind down, in the second one, the integral rapidly winds down and in the 3rd one, the integral didn't wind up at all.

Setting the parameter `MPC_XY_VEL_ARW` < -0.5 disables ARW. Between -0.5 and 0 the legacy ARW is used and > 0, the new strategy is used and the value corresponds to the ARW gain. A good starting value is 3.0.

**Describe problem solved by the proposed pull request**
Integral windup is a common problem in PID loops and prevents increasing the integral gain without increasing the overshoot. The strategy used so far was sufficient to prevent large windups but not good enough to be able to set really aggressive gains.

**Describe your preferred solution**
The new ARW introduced by this PR is known as "Tracking ARW" and is a simple feedback loop that uses the integral to track the saturation.

**Describe possible alternatives**
Another strategy is to set the integral such that it always unsaturates the output. The drawback of that method is that it is sensitive to spikes. 

**Additional context**
I'll add something in the user guide.
- https://ac.els-cdn.com/S1474667017518650/1-s2.0-S1474667017518650-main.pdf?_tid=ed135642-d163-4aa6-98e7-4a912a5a87a4&acdnat=1538986751_bb7f14c5bd5e1a5803f50659631e671d

 FYI @dagar , @LorenzMeier , @thomasgubler

### CODE REVIEW

**Comment by dagar** [comment]
CI failure is a trivial style issue. Can you fix and rebase on master?
image

http://ci.px4.io:8080/blue/organizations/jenkins/PX4%2FFirmware/detail/dev-velocity-arw/4/pipeline

**Comment by bresch** [comment]
Done.
Btw, do we let the choice to the user to use the legacy ARW or do we remove completely the code? I still have it in that PR to be able to compare.

**Comment by dagar** [comment]
> do we let the choice to the user to use the legacy ARW or do we remove completely the code?

If you already consider it safe to enable the new method by default my initial preference is to drop the old. Realistically we're only going to test the main path, which means it wouldn't be that difficult for someone to unintentionally break the old code.

 Any opinion @RomanBapst/@Stifael/@MaEtUgR?

**Comment by santiago3dr** [comment]
Couple flights on a Pixracer (V4)
steady flight; left MPC_XY_VEL_ARW set to 3
https://logs.px4.io/plot_app?log=34d58da4-48e8-411d-ab27-78477f5d0e82
https://logs.px4.io/plot_app?log=1196eebf-46cb-4289-9d15-78fb2f7c5005

similar flights on current master 6c34407 for comparison:
https://review.px4.io/plot_app?log=3d63d518-4b0d-481d-8f45-32317e1b2397
https://review.px4.io/plot_app?log=09c273d0-64ea-4804-8dd8-a191ad7c74c2

File: src/modules/mc_pos_control/PositionControl.cpp [hunk 1]
```diff
@@ -306,25 +306,41 @@ void PositionControl::_velocityController(const float &dt)
...
-			if (integral_mag_NE > 0.0f && integral_mag_NE > thrust_max_NE) {
-				_thr_int(0) = _thr_int(0) / integral_mag_NE * thrust_max_NE;
-				_thr_int(1) = _thr_int(1) / integral_mag_NE * thrust_max_NE;
+		if (MPC_XY_VEL_ARW.get() > FLT_EPSILON) { // Standard anti-reset windup
+			Vector2f vel_err_lim;
+			vel_err_lim(0) = vel_err(0) - (thrust_desired_NE(0) - _thr_sp(0)) * (MPC_XY_VEL_ARW.get());
+			vel_err_lim(1) = vel_err(1) - (thrust_desired_NE(1) - _thr_sp(1)) * (MPC_XY_VEL_ARW.get());
+
+			// Update integral
+			_thr_int(0) += MPC_Z_VEL_I.get() * vel_err_lim(0) * dt;
```
Comment (Stifael) [inline src/modules/mc_pos_control/PositionControl.cpp]: Why do you use the z-integral term?

**Comment by Stifael** [comment]
I have a couple comments:
In our two-loop P-PID controllers, the only part of the controller that can close the steady-state error in position is the velocity integral term. This means that the effect and advantage of the I-term is primarily visible if there is some sort of disturbance. In the Z-direction, that would be gravity, and in the xy-direction mainly wind (besides air-resistance of course). On the other hand, the integrator has a negative effect durin step inputs. Therefore, for testing I also would add plots where the vehicle is not just doing a step input, because for step inputs in xy the best result is achieved if the in the integrator is turned off entirely.
I also was wondering why you only changed the integrator logic in xy and not in z. In fact, in most cases the integrator in xy can be dropped, whereas in the z-direction it continuously corrects for the hover-thrust value.

**Comment by Stifael** [comment]
@santiago3dr during testing, make sure that position/velocity control related smoothing is turned off. You can do that by settin MPC_POS_MODE to 0 https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/mc_pos_control_params.c#L629.

**Comment by santiago3dr** [comment]
@Stifael couple flights with MPC_POS_MODE set to 0
https://logs.px4.io/plot_app?log=83d4ec30-00fb-4e05-b569-c4d5fe8acdd4
https://logs.px4.io/plot_app?log=1a7d9f78-777a-4cac-bfb3-6b5bbd082d8d

**Comment by dagar** [comment]
Replaced by https://github.com/PX4/Firmware/pull/10708

**Pull request closed without merging**